Fix CME sub-specialties

stvstnfrd opened the pull request:
While the model doesn't enforce choices for the sub-specialty,
the form apparently does. We had been arbitrarily defaulting the
sub-specialty input dropdown with only the blank entry `-----`, since we
overwrite after the specialty is updated. A side-effect of this was that
the form backend thought any non-empty sub-specialty was invalid.

Valid sub-specialty values are now allowed.

stvstnfrd commented:
Please review: @dcadams @caseylitton @caesar2164 @kluo 

NB: I distinctly remember testing this functionality manually during development, so I'm still perplexed about what really changed here.

stvstnfrd commented:
NB2: You can verify the behavior by running the tests [1]
with just the `test_form.py` commit; tests will fail.
Then pull in the `form.py` change; tests will pass.

[1] `paver test_system -s lms -t openedx/stanford/djangoapps/register_cme/tests/test_form.py`

caseylitton commented:
+1

kluo commented on openedx/stanford/djangoapps/register_cme/forms.py:
```diff
@@ -7,6 +7,17 @@
+    choices = [
+        ('', '---------'),
+    ]
+    for specialty, sub_specialties in SUB_SPECIALTY_CHOICES.items():
```
Can we just use `SUB_SPECIALTY_CHOICES.values()` since `specialty` is not used?

stvstnfrd commented:
I added a new commit [1] (which I'll later squash),
per @kluo's suggestion.
I also refactored that section so that the list of all
sub-section choices is cached at compile time,
instead of being regenerated upon each page load.

I had wanted to include the caching initially,
but I struggled with the comprehension syntax;
I managed to work it out now.

[1] ebc25fe64f7efa41ad7e322a583429e88ef2aaa3

kluo commented:
Cool, so just changing the choices from calling a helper function to a computed constant affects the caching behavior? Didn't see Django docs mention this

stvstnfrd commented:
From chat:

> [13:52] it’s nothing specific to django, just the way python evaluates the code…

> [13:54] the first time that file is imported/executed anyway, the
>         interpreter starts working through the file, line by line. so
>         w/ the new code, the first time, it sets a “constant”
>         `ALL_SUB_SPECIALTY_CHOICES`, then that constant is used.

> [13:55] but when a helper is used instead of the constant, it means
>         that function gets called again each time you instantiate a
>         new `ExtraInfoForm` object
